### PR TITLE
Change macro expansion mangling to avoid request cycles

### PR DESCRIPTION
**Pull request by slavapestov**
Fixes rdar://127078338.

### CODE REVIEW

File: lib/AST/Expr.cpp [hunk 1]
```diff
@@ -1934,7 +1934,12 @@ unsigned AbstractClosureExpr::getDiscriminator() const {
         Bits.AbstractClosureExpr.Discriminator = discriminator;
   }
 
-  assert(getRawDiscriminator() != InvalidDiscriminator);
+  if (getRawDiscriminator() == InvalidDiscriminator) {
+    llvm::errs() << "Closure does not have an assigned discriminator:\n";
+    dump(llvm::errs());
+    abort();
```
Comment (DougGregor) [inline lib/AST/Expr.cpp:1940]: Can we please produce a diagnostic instead of aborting? This kind of bug shouldn't take down a process.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: test/Macros/skip_non_exportable_decls.swift [hunk 2]
```diff
@@ -10,8 +10,10 @@
...
 
-// CHECK: sil @$s9MacroUser03$s9A70User33_B2D49A1BE4DC7AF5CC327EB8EE2214BDLl14anonymousTypesfMf_4namefMu_C5helloSSyF : $@convention(method) (@guaranteed $s9MacroUser33_B2D49A1BE4DC7AF5CC327EB8EE2214BDLl14anonymousTypesfMf_4namefMu_) -> @owned String {
+// CHECK: sil @$s9MacroUser03$s9A118User0036skip_non_exportable_declsswift_tjBBlfMX14_0_33_B2D49A1BE4DC7AF5CC327EB8EE2214BDLl14anonymousTypesfMf_4namefMu_C5helloSSyF : $@convention(method) (@guaranteed $s9MacroUser0036skip_non_exportable_declsswift_tjBBlfMX14_0_33_B2D49A1BE4DC7AF5CC327EB8EE2214BDLl14anonymousTypesfMf_4namefMu_) -> @owned String {
+
 #anonymousTypes(public: true) { "hello" }
 
-// CHECK-NOT: s9MacroUser03$s9A71User33_B2D49A1BE4DC7AF5CC327EB8EE2214BDLl14anonymousTypesfMf_4namefMu0_O5helloSSyF
 #anonymousTypes(public: false) { "goodbye" }
+
+// FIXME: This test
```
Comment (ahoppen) [inline test/Macros/skip_non_exportable_decls.swift:19]: Drive-by comment: Is this a leftover comment?

**Comment by slavapestov** [comment]
https://github.com/apple/llvm-project/pull/8893
@swift-ci Please smoke test macOS

File: test/Macros/skip_non_exportable_decls.swift (hunk 2 above)
Comment (slavapestov) [reply test/Macros/skip_non_exportable_decls.swift:19]: Nope, the test is broken or there's a bug; I noticed that changing 'public' to true has no effect on the SIL output, which I think is wrong.

Comment (ahoppen) [reply test/Macros/skip_non_exportable_decls.swift:19]: Ah, I see what you by the FIXME now with your comment. Do you think it would be valuable to include more information about what in the test needs fixing (essentially putting your GitHub comment in the source code)? I’m just thinking that somebody reading this FIXME in a year or two might have a hard time understanding what it was referring to.

Comment (slavapestov) [reply test/Macros/skip_non_exportable_decls.swift]: Done

**Comment by slavapestov** [comment]
https://github.com/apple/llvm-project/pull/8893
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
https://github.com/apple/llvm-project/pull/8894
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
https://github.com/apple/llvm-project/pull/8898
@swift-ci Please smoke test
